Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/features/settings/ui/AchievementProgressAdapter.kt

```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+package org.dolphinemu.dolphinemu.features.settings.ui
+
+import android.view.LayoutInflater
+import android.view.ViewGroup
+import androidx.recyclerview.widget.RecyclerView
+import org.dolphinemu.dolphinemu.databinding.ListItemAchievementProgressBinding
+import org.dolphinemu.dolphinemu.databinding.ListItemHeaderBinding
+import org.dolphinemu.dolphinemu.features.settings.model.AchievementProgressViewModel
+import org.dolphinemu.dolphinemu.features.settings.ui.viewholder.AchievementProgressItemViewHolder
+import org.dolphinemu.dolphinemu.features.settings.ui.viewholder.AchievementProgressViewHolder
+import org.dolphinemu.dolphinemu.features.settings.ui.viewholder.AchievementHeaderViewHolder
+
+class AchievementProgressAdapter(
+    private val activity: SettingsActivity,
+    private val viewModel: AchievementProgressViewModel
+) : RecyclerView.Adapter<AchievementProgressItemViewHolder>() {
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AchievementProgressItemViewHolder {
+        val inflater = LayoutInflater.from(parent.context)
+        return when (viewType) {
+            AchievementProgressItem.Companion.TYPE_ACHIEVEMENT -> {
+                val listItemAchievementProgressBinding = ListItemAchievementProgressBinding.inflate(inflater, parent, false)
+                AchievementProgressViewHolder(listItemAchievementProgressBinding)
+            }
+            AchievementProgressItem.Companion.TYPE_HEADER -> {
+                val listItemHeaderBinding = ListItemHeaderBinding.inflate(inflater, parent, false)
+                AchievementHeaderViewHolder(listItemHeaderBinding)
+            }
+            else -> throw UnsupportedOperationException()
+        }
+    }
+
+    override fun onBindViewHolder(holder: AchievementProgressItemViewHolder, position: Int) {
+        holder.bind(activity, getItemAt(position), position)
+    }
+
+    override fun getItemCount(): Int {
+        return viewModel.buckets.size + viewModel.buckets.sumOf { bucket -> bucket.achievements.size }
+    }
+
+    override fun getItemViewType(position: Int): Int {
+        return getItemAt(position).type
+    }
+
+    private fun getItemAt(position: Int): AchievementProgressItem {
+        var itemPosition = position
```

Comment:
You're not decrementing `itemPosition` or anything along those lones after going through each bucket, so this will only work correctly for the first bucket.